src/GCDNotifier.cs

```diff
@@ -0,0 +1,308 @@
+using GCDTracker.UI;
+using System.Collections.Generic;
+using Dalamud.Interface.Animation;
+using Dalamud.Interface.Animation.EasingFunctions;
+using System.Numerics;
+using System.Linq;
+using System;
+
+namespace GCDTracker {
+
+    public enum EventType {
+        FlyOutAlert,
+        BarColorPulse,
+        BarWidthPulse,
+        BarHeightPulse,
+        BarBackground,
+        WheelPulse,
+        FloatingTriangle,
+        None
+    }
+
+    public enum EventCause {
+        Slidecast,
+        Queuelock,
+        Clipped,
+        ABC,
+        None
+    }
+
+    public class Alert
```

Comment:
I have a big change I would make here. I know you don't have time to tackle this, but I would love to get your opinion on it.

The biggest refactor I would make is changing Alert to an abstract class and making `FlyOutAlert, PulseAlert and FloatingAlert` inherit from it. This way we can give a lot more responsibility to the alert itself, such as:
 - Passing them a reference to the Bar and Wheel so they can change properties such as Color/Size by themselves.
 - Allowing them to draw their own config options on the Configuration screen on something like `DrawConfig(bar/wheel)`
 - Provide an `IsFinished()` that allows the AlertManager to remove them once finished. This way we can also keep the Alert as an active alert and bypass the MarkAlert/CheckAlert.
 - Animating properties by changing the HandleAlert in AlertManager to calling Update() on existing alerts.
 - Drawing themselves by making an `AlertManager.Draw()` that triggers the draw for existing alerts. 

I think the only problem this causes is that we need a way to handle priorities for things like ABC/Clip, which we can do by either using a prioirty list or trusting the rest of the code to only send one of the events.

Reply:
ABC/Clip should both already be fire once already.  I added some checks to make sure that they don’t enqueue a second alert but they really aren’t necessary.  More of a “better safe than sorry” check.  ABC already checks to see if a clip triggered before setting itself.  Said another way clip always takes precedence over ABC.  This functionality will always be necessary for the bar/wheel background color, and I think that should always stay in display/helper anyway so we can safely trust only having one.

Only other comment is to watch out for double alerts when someone has the bar and the wheel both active.  I set it up to try to make sure that bar and wheel related alerts would only get processed when PluginUI was in the right place to handle those alerts. 

Trying to do it async lead me down a whole adventure of null checks and alerts showing up in the wrong place because we draw them relative to their parent object, and depending on the current state of PluginUI we get very different outcomes.

so that’s how I got to where I did.

I would tend to agree that if all of those problems could be solved, the vision you outlined above is much more versatile and feature-rich than what I’ve done.


It lead to unique possibilities like letting the user set a specific time / condition they want a pulse or alert to happen at entirely unrelated to the other feature we have.


I’d add that if we are thinking of making big changes we should look at decoupling the bar and wheel progress from the gcdTotal/castTotal and just use the system time to do it ourselves.  That would let us avoid a whole host of issues due to the way the gcdTotal is sometimes set to 0 by the game even though an action is happening.

> Drawing themselves by making an AlertManager.Draw() that triggers the draw for existing alerts.

This would be great for the abc/clip alerts, maybe not as useful for the bar pulses and wheel pulse.  We can’t supersede the actual drawing of the bar/wheel is active unless we also want to take over all of the logic that is going on to draw the actual bar/wheel.  There is always going to need to be some coordination between what’s going on in GCDDisplay and GCDNotifier.


Maybe neither of them should invoke draw directly.  Maybe they should both update state of a shared class that invokes draw when one of them instructs it to?